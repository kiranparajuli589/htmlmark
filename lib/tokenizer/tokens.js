const TOKENS = {
  newLine: "new-line",
  paragraph: "paragraph",
  codeBlock: "code-block",
  comment: "comment",
  image: "image",
  quote: "quote",
  bold: "bold",
  italic: "italic",
  underline: "underline",
  strikeThrough: "strike-through",
  code: "code",
  link: "link",
  listItem: "list-item",
  countItem: "count-item",
  heading: "heading",
  checkBox: "check-box",
  checkBoxChecked: "check-box-checked",
  text: "text"
}

module.exports = TOKENS
