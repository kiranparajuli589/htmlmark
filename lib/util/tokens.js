const TOKENS = {
  NEW_LINE: "new-line",
  PARAGRAPH: "paragraph",
  CODE_BLOCK: "code-block",
  COMMENT: "comment",
  IMAGE: "image",
  QUOTE: "quote",
  BOLD: "bold",
  ITALIC: "italic",
  UNDERLINE: "underline",
  STRIKE_THROUGH: "strike-through",
  CODE: "code",
  LINK: "link",
  LIST_ITEM: "list-item",
  COUNT_ITEM: "count-item",
  HEADING: "heading",
  CHECKBOX: "check-box",
  CHECKBOX_CHECKED: "check-box-checked",
  TEXT: "text",
  HR_LINE: "hr-line",
  TABLE: "table",
}

module.exports = TOKENS
