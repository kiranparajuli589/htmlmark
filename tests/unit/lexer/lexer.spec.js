import { Lexer } from "../../../lib/lexer/index.js"
import { TOKENS } from "../../../lib/util/tokens.js"
import { commonTokensList, LinesToEscape } from "../../fixtures/commTokens.js"


describe("lexer", () => {
	describe("hr line", () => {
		it.each([
			"---",
			"***"
		])("should parse the hr line: %s", (line) => {
			const lines = [
				line
			]
			const lexer = new Lexer(lines)
			const tokens = lexer.run()
			expect(tokens[0].type).toBe(TOKENS.HR_LINE)
		})
		it("should merge multiple consecutive hr lines into one", () => {
			const lines = [
				"---",
				"---"
			]
			const lexer = new Lexer(lines)
			const tokens = lexer.run()
			expect(tokens.length).toBe(1)
			expect(tokens[0].type).toBe(TOKENS.HR_LINE)
		})
	})
	describe("common tokens", () => {
		it.each(commonTokensList)("should parse the common tokens: '%s'", (line) => {
			const lines = [line]
			const lexer = new Lexer(lines)
			const tokens = lexer.run()
			expect(tokens).toMatchSnapshot()
		})
		it.each(commonTokensList)("should parse the common tokens with indentation", (line) => {
			const indent = " ".repeat(4)
			const lines = [indent + line]
			const lexer = new Lexer(lines)
			const tokens = lexer.run()
			expect(tokens[0].indent).toBe(4)
		})
		it("should allow underlined heading", () => {
			const lines = [
				"Heading 1",
				"======",
				"Heading 2",
				"------",
				"Normal text"
			]
			const lexer = new Lexer(lines)
			const tokens = lexer.run()
			expect(tokens).toMatchSnapshot()
		})
	})
	describe("escaping", () => {
		it("should escape everything inside a code block", () => {
			const lines = [
				"```",
				...LinesToEscape,
				"```"
			]
			const lexer = new Lexer(lines)
			const tokens = lexer.run()
			expect(tokens).toMatchSnapshot()
		})
		it("should escape everything excepts tags for other tokens", () => {
			const lines = [
				"# &Ampersand",
				"- 'SomeQuote'",
				"> Some \" comment",
				"<bold>bold</bold>"
			]
			const lexer = new Lexer(lines)
			const tokens = lexer.run()
			expect(tokens).toMatchSnapshot()
		})
	})
})
