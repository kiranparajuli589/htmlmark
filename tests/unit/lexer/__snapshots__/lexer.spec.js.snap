// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lexer common tokens should allow underlined heading 1`] = `
Array [
  Object {
    "indent": 0,
    "level": 1,
    "raw": "Heading 1
======",
    "setext": true,
    "tokens": Array [
      Object {
        "raw": "Heading 1",
        "type": "text",
        "value": "Heading 1",
      },
    ],
    "type": "heading",
    "value": "Heading 1",
  },
  Object {
    "indent": 0,
    "level": 2,
    "raw": "Heading 2
------",
    "setext": true,
    "tokens": Array [
      Object {
        "raw": "Heading 2",
        "type": "text",
        "value": "Heading 2",
      },
    ],
    "type": "heading",
    "value": "Heading 2",
  },
  Object {
    "indent": 0,
    "raw": "Normal text",
    "tokens": Array [
      Object {
        "raw": "Normal text",
        "type": "text",
        "value": "Normal text",
      },
    ],
    "type": "paragraph",
    "value": "Normal text",
  },
]
`;

exports[`lexer common tokens should parse the common tokens: '![alt text](image.png)' 1`] = `
Array [
  Object {
    "alt": "alt text",
    "indent": 0,
    "raw": "![alt text](image.png)",
    "type": "image",
    "url": "image.png",
  },
]
`;

exports[`lexer common tokens should parse the common tokens: '# Heading One Text' 1`] = `
Array [
  Object {
    "indent": 0,
    "level": 1,
    "raw": "# Heading One Text",
    "setext": false,
    "tokens": Array [
      Object {
        "raw": "Heading One Text",
        "type": "text",
        "value": "Heading One Text",
      },
    ],
    "type": "heading",
    "value": "Heading One Text",
  },
]
`;

exports[`lexer common tokens should parse the common tokens: '## Heading Two Text' 1`] = `
Array [
  Object {
    "indent": 0,
    "level": 2,
    "raw": "## Heading Two Text",
    "setext": false,
    "tokens": Array [
      Object {
        "raw": "Heading Two Text",
        "type": "text",
        "value": "Heading Two Text",
      },
    ],
    "type": "heading",
    "value": "Heading Two Text",
  },
]
`;

exports[`lexer common tokens should parse the common tokens: '### Heading Three Text' 1`] = `
Array [
  Object {
    "indent": 0,
    "level": 3,
    "raw": "### Heading Three Text",
    "setext": false,
    "tokens": Array [
      Object {
        "raw": "Heading Three Text",
        "type": "text",
        "value": "Heading Three Text",
      },
    ],
    "type": "heading",
    "value": "Heading Three Text",
  },
]
`;

exports[`lexer common tokens should parse the common tokens: '#### Heading Four Text' 1`] = `
Array [
  Object {
    "indent": 0,
    "level": 4,
    "raw": "#### Heading Four Text",
    "setext": false,
    "tokens": Array [
      Object {
        "raw": "Heading Four Text",
        "type": "text",
        "value": "Heading Four Text",
      },
    ],
    "type": "heading",
    "value": "Heading Four Text",
  },
]
`;

exports[`lexer common tokens should parse the common tokens: '##### Heading Five Text' 1`] = `
Array [
  Object {
    "indent": 0,
    "level": 5,
    "raw": "##### Heading Five Text",
    "setext": false,
    "tokens": Array [
      Object {
        "raw": "Heading Five Text",
        "type": "text",
        "value": "Heading Five Text",
      },
    ],
    "type": "heading",
    "value": "Heading Five Text",
  },
]
`;

exports[`lexer common tokens should parse the common tokens: '###### Heading Six Text' 1`] = `
Array [
  Object {
    "indent": 0,
    "level": 6,
    "raw": "###### Heading Six Text",
    "setext": false,
    "tokens": Array [
      Object {
        "raw": "Heading Six Text",
        "type": "text",
        "value": "Heading Six Text",
      },
    ],
    "type": "heading",
    "value": "Heading Six Text",
  },
]
`;

exports[`lexer common tokens should parse the common tokens: '<!-- comment item text -->' 1`] = `
Array [
  Object {
    "indent": 0,
    "raw": "<!-- comment item text -->",
    "type": "comment",
    "value": "comment item text",
  },
]
`;

exports[`lexer escaping should escape everything excepts tags for other tokens 1`] = `
Array [
  Object {
    "indent": 0,
    "level": 1,
    "raw": "# &Ampersand",
    "setext": false,
    "tokens": Array [
      Object {
        "raw": "&amp;Ampersand",
        "type": "text",
        "value": "&amp;amp;Ampersand",
      },
    ],
    "type": "heading",
    "value": "&amp;Ampersand",
  },
  Object {
    "indent": 0,
    "items": Array [
      Object {
        "checked": null,
        "count": null,
        "raw": "- 'SomeQuote'",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "'SomeQuote'",
            "tokens": Array [
              Object {
                "raw": "'SomeQuote'",
                "type": "text",
                "value": "&#39;SomeQuote&#39;",
              },
            ],
            "type": "paragraph",
            "value": "'SomeQuote'",
          },
        ],
        "type": "list-item",
      },
    ],
    "meta": Object {
      "checklist": false,
      "identifier": "-",
      "ordered": false,
    },
    "raw": "- 'SomeQuote'",
    "type": "list",
  },
  Object {
    "depth": 1,
    "indent": 0,
    "raw": "> Some \\" comment
<bold>bold</bold>",
    "tokens": Array [
      Object {
        "indent": 0,
        "raw": " Some \\" comment
<bold>bold</bold>",
        "tokens": Array [
          Object {
            "raw": " Some \\" comment <bold>bold</bold>",
            "type": "text",
            "value": " Some &quot; comment <bold>bold</bold>",
          },
        ],
        "type": "paragraph",
        "value": " Some \\" comment <bold>bold</bold>",
      },
    ],
    "type": "quote",
  },
]
`;

exports[`lexer escaping should escape everything inside a code block 1`] = `
Array [
  Object {
    "indent": 0,
    "language": null,
    "raw": "\`\`\`
<div>
</div>
ab<cd
efg>
efg<
>>>
<<<<
& ampersand
'single quotes'
\\"double quotes\\"
\`\`\`",
    "type": "code-block",
    "value": "&lt;div&gt;
&lt;/div&gt;
ab&lt;cd
efg&gt;
efg&lt;
&gt;&gt;&gt;
&lt;&lt;&lt;&lt;
&amp; ampersand
&#39;single quotes&#39;
&quot;double quotes&quot;",
  },
]
`;

exports[`lexer link variables should allow the link variables 1`] = `
Array [
  Object {
    "indent": 0,
    "raw": "If you want to search for something on the web, [google] is a nice place.",
    "tokens": Array [
      Object {
        "raw": "If you want to search for something on the web, ",
        "type": "text",
        "value": "If you want to search for something on the web, ",
      },
      Object {
        "raw": "[google]",
        "tokens": Object {
          "href": "https://www.google.com/",
          "title": Object {
            "raw": "google",
            "tokens": Array [
              Object {
                "raw": "google",
                "type": "text",
                "value": "google",
              },
            ],
          },
          "tooltip": undefined,
        },
        "type": "link",
      },
      Object {
        "raw": " is a nice place.",
        "type": "text",
        "value": " is a nice place.",
      },
    ],
    "type": "paragraph",
    "value": "If you want to search for something on the web, [google] is a nice place.",
  },
  Object {
    "type": "new-line",
  },
  Object {
    "indent": 0,
    "raw": "You can also use [google] to host some stuff and make a great app!",
    "tokens": Array [
      Object {
        "raw": "You can also use ",
        "type": "text",
        "value": "You can also use ",
      },
      Object {
        "raw": "[google]",
        "tokens": Object {
          "href": "https://www.google.com/",
          "title": Object {
            "raw": "google",
            "tokens": Array [
              Object {
                "raw": "google",
                "type": "text",
                "value": "google",
              },
            ],
          },
          "tooltip": undefined,
        },
        "type": "link",
      },
      Object {
        "raw": " to host some stuff and make a great app!",
        "type": "text",
        "value": " to host some stuff and make a great app!",
      },
    ],
    "type": "paragraph",
    "value": "You can also use [google] to host some stuff and make a great app!",
  },
]
`;
